File: python/minecraft_bot/protocol/v763/packets/play/clientbound/stop_sound.py
"""Packet `stop_sound` (play/clientbound, id 0x63).

Stops a currently-playing sound. ``flags`` is a bitfield::

    0x01  source provided
    0x02  sound name provided

Both fields are optional and present only when their flag bit is set.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import Optional

from minecraft_bot.codec import Reader, Writer, string, varint

PACKET_ID = 0x63


@dataclass(frozen=True, slots=True)
class StopSound:
    flags: int                # i8 bitfield
    source: Optional[int]     # varint, present iff (flags & 0x01)
    sound: Optional[str]      # identifier, present iff (flags & 0x02)


def decode(reader: Reader) -> StopSound:
    flags, = struct.unpack(">b", reader.read(1))
    src: Optional[int] = None
    snd: Optional[str] = None
    if flags & 0x01:
        src = varint.read(reader)
    if flags & 0x02:
        snd = string.read(reader)
    return StopSound(flags=flags, source=src, sound=snd)


def encode(packet: StopSound, writer: Writer) -> None:
    writer.write(struct.pack(">b", packet.flags))
    if packet.flags & 0x01:
        if packet.source is None:
            from minecraft_bot.errors import ValueOutOfRange
            raise ValueOutOfRange("stop_sound.source", packet.source)
        varint.write(packet.source, writer)
    if packet.flags & 0x02:
        if packet.sound is None:
            from minecraft_bot.errors import ValueOutOfRange
            raise ValueOutOfRange("stop_sound.sound", packet.sound)
        string.write(packet.sound, writer)

File: rust/src/protocol/v763/packets/play/clientbound/stop_sound.rs
//! Packet `stop_sound` (play/clientbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot, string_codec as string, varint,
    varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x63;

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct StopSound {
    pub flags: i8,
    pub source: Option<i32>,
    pub sound: Option<String>,
}

impl StopSound {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let flags = reader.read_exact(1)?[0] as i8;
        let source = if flags & 0x01 != 0 {
            Some(varint::read(reader)?)
        } else {
            None
        };
        let sound = if flags & 0x02 != 0 {
            Some(string::read(reader)?)
        } else {
            None
        };
        Ok(Self {
            flags,
            source,
            sound,
        })
    }
}

impl ClientboundPacket for StopSound {
    fn state(&self) -> ConnectionState {
        ConnectionState::Play
    }
    fn packet_id(&self) -> i32 {
        PACKET_ID
    }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        writer.write_all(&[self.flags as u8])?;
        if let Some(s) = self.source {
            varint::write(s, writer)?;
        }
        if let Some(s) = &self.sound {
            string::write(s, writer)?;
        }
        Ok(())
    }
}
